RBAC API for Metrics Access

ckandag opened the pull request:
This is an initial pass at FinerGrained RBAC API for Metrics Access Related to: https://issues.redhat.com/browse/ACM-2420

Signed-off-by: Chaitanya Kandagatla <ckandaga@redhat.com>

ckandag commented:
Sonar cloud is isnt resolved. Would appreciate any review comments while I work on setting up the checks correctly

mprahl commented on OWNERS:
```diff
@@ -0,0 +1,16 @@
+approvers:
+- ChunxiAlexLuo
```
Could you please remove Alex from this file since he's no longer on the team?

gparvin replied:
Instead of adding Alex to the Owners, let's add Christine: `yiraeChristineKim`

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+	verb: "metrics/",
+}
+
+// APIEngine is the rbac api for custom/finer grained RBAC, it should preferably be instantiated through
```
Since the struct fields are private.
```suggestion
// APIEngine is the rbac api for custom/finer grained RBAC, it must be instantiated through
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+	kubeClient kubernetes.Interface
+}
+
+// kubeConfig  - Kubernetes Cluster configuration
```
Could you please use the names (i.e. `kubeConfig` => `kConfig`) that are used in the function signature?

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+// if not, any configured kubernetes client on the RbacEgine is returned
+func (r *APIEngine) getKubeClientForUser(userToken string) (kubernetes.Interface, error) {
+	if len(userToken) != 0 && r.kubeConfig != nil {
+		// ?? maybe I should make a copy of the kubeConfig everytime instead of setting it on the same copy
```
Good point. A copy would be best.

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+	} else if r.kubeConfig == nil {
+		return r.kubeClient, nil
+	} else {
+		return nil, fmt.Errorf("failed to evaluate the acls for User")
```
Could you please provide a more informative error here?

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+// kubeClient  - Client to talk to a Kubernetes cluster
+// Either or both Config and Client to connect to the Kubernetes cluster can be set.
+// throws an error when neither is set.
+func NewAPIEngine(kConfig *rest.Config, kClient kubernetes.Interface) (*APIEngine, error) {
```
Can you make a copy of `kConfig` since this is a pointer? I wouldn't want the caller to be able to modify `kConfig` outside of `APIEngine` and have it affect the behavior of `APIEngine`.

ckandag replied:
@mprahl -  I am not so convinced about this

At the other location (getKubeClientForUser), because the user tokens may unintentionally overwrite each other, we would need to make a copy

But here, may be the caller changes the settings and wants them to take effect everywhere.  This isn't a service , just a library API, may be fair to expect the changes to take everywhere.

mprahl replied:
I think that would be undefined behavior if you change the config after instantiating the struct.

I would like to follow what client-go does for `NewForConfig`:
https://github.com/kubernetes/client-go/blob/babfe96631905312244f34500d2fd1a8a10b186c/kubernetes/clientset.go#L453

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+// Either or both Config and Client to connect to the Kubernetes cluster can be set.
+// throws an error when neither is set.
+func NewAPIEngine(kConfig *rest.Config, kClient kubernetes.Interface) (*APIEngine, error) {
+	if kClient == nil && kConfig == nil {
```
Optional: I'd prefer if this only accepted one of the options since if both are provided, it's not super clear which is used where. This would also make `getKubeClientForUser` a bit simpler.

ckandag replied:
I would prefer both  to be able to support both.  I actually started out with just kubeconfig but realized there will be cases where consumers only have a kubeclient or bearer token is not available at all , like in the case of the tests etc.

I can  improve the  documentation and error msgs in the method.

If you have suggestions around how to better implement this while still supporting both options, will appreciate it

mprahl replied:
@ckandag I was trying to say you could error if both are provided but still support both.

ckandag replied:
ah ok

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+	}
+}
+
+// this API call returns the namespaces allowed for  metrics access for the  given list of managedcluster (or all)
```
This is the common pattern in Go documentation.
```suggestion
// GetMetricsAccess returns the namespaces allowed for metrics access for the  given list of managedcluster (or all)
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+	return metricsAccessResults, nil
+}
+
+// returns all Acls for a given resource type
```
```suggestion
// GetResourceAccess returns all Acls for a given resource type
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+// returns all Acls for a given resource type
+// if resource names is non-empty, acls are filtered for only the given list of resourcenames ,
+// else if resource names is empty,acls for resources of the given type that user has access to are returned
+// namespace is used for namespace-scoped resources, for cluster-scoped resources is can be left empty
```
```suggestion
// namespace is used for namespace-scoped resources, for cluster-scoped resources should be left empty
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+	return resultList
+}
+
+// makes a selfsubjectrulesreview call , the call requires a namespace to be set
```
```suggestion
// MakeSubjectAccessRulesReviewForUser makes a selfsubjectrulesreview call , the call requires a namespace to be set
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+}
+
+// makes a selfsubjectrulesreview call , the call requires a namespace to be set
+// call returns the list of the allowed acls for the user
```
This isn't quite flowing from the previous comment line.

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+// namespace is used for namespace-scoped resources, for cluster-scoped resources is can be left empty
+// for a given cluster, if no  acls are configured, an empty list is returned for it in the results
+
+func GetResourceAccess(
```
Can you document that `namespace` defaults to `default` if it isn't specified? That seems to be the behavior in `MakeSubjectAccessRulesReviewForUser`.

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+
+	sarrStatus := response.Status
+
+	klog.V(2).Infof("Is ssrr incomplete : %v", namespace)
```
```suggestion
	klog.V(2).Infof("Is sarr incomplete : %v", namespace)
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+) ([]authorizationv1.ResourceRule, error) {
+	klog.V(2).Infof("Make Subject Access Rules Review for Namespace %s", namespace)
+
+	if len(namespace) == 0 {
```
Optional: This is the more common approach in Go
```suggestion
	if namespace == "" {
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+// a convenience method for building a list with unique entries
+// specified items are added to the given  slice  of items if not already in it
+
+func addUniqueItems(itemlist []string, itemsToAdd ...string) []string {
```
Optional: The better approach would be to use a set (i.e. `map[string]bool`) and then convert it to a slice after all the results are added to the set.

dhaiducek commented:
Sure! I can take a look!

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+		// filter the rules by the given ApiGroup(or *) and Resource(or *))
+		ruleMatchesAPIGroup := (slices.Contains(rule.APIGroups, gr.Group) || slices.Contains(rule.APIGroups, "*"))
+		ruleMatchesResource := (slices.Contains(rule.Resources, gr.Resource) || slices.Contains(rule.Resources, "*"))
+		if ruleMatchesAPIGroup && ruleMatchesResource {
```
Optional: To reduce the indentation and improve readability.
```suggestion
		if !(ruleMatchesAPIGroup && ruleMatchesResource) {
			continue
		}
```

mprahl commented on pkg/rbac/rbac_test.go:
```diff
@@ -0,0 +1,182 @@
+			},
+			nil,
+		},
+		{ // user-purple is in both blue-admin & red-admin usergroup,
```
This test fails for me locally. I think the ordering is not deterministic so your equality checks may need to use sets.

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+	klog.V(2).Infof("Is ssrr incomplete : %v", namespace)
+
+	if sarrStatus.Incomplete || len(sarrStatus.EvaluationError) != 0 {
+		return nil, fmt.Errorf("failed to evaluate the acls for User")
```
Could you include why it failed in this error?

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+// for a given cluster,
+//			if all access is configured for the user, i.e the verb "*",  list of the namespaces is set to *
+
+func (r *APIEngine) GetMetricsAccess(userToken string, clusters ...string) (map[string][]string, error) {
```
Does this also need to take into account that if the user has `get` access on the `ManagedCluster` object they would have access to all namespaces for metrics? I don't recall where we landed in that discussion.

ckandag replied:
@mprahl  So, I guess we could but  today observability does not require User to have a `get` on ManagedCluster. Instead, requires/checks for user to have access to the MC's Hub namespace( the old day before the ManagedCluster CR was introduced).   Need to look into how backward compatibility will be supported.

I will take a separate bug/issue to address this . May not resolve in this PR

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+		for _, acl := range clusteracls {
+			// filter verbs that start with metrics/ and grab the namespace
+			// if verb is set to *, set namespace to * to indicate access to all namespaces
+			if strings.HasPrefix(acl, MetricsACLConfig.verb) {
```
Would the verb `metrics` be a valid option to mean metrics access to all namespaces on the managed cluster? This code doesn't account for that.

ckandag replied:
Not having any  specific "metrics/" set  was discussed, 
i.e  if user has access to the ManagedCluster  but no specific metrics/<namespaces> access is set, then for backward compatibility its considered access to all.

I do not recall discussing "metrics" to mean all namespaces,   but we could support that
 I feel metrics/* probably is better , its more intentional and clear.  The current code would support that case.

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+		klog.V(2).Infof("clustername [%s] acls[%s]\n", clustername, clusteracls)
+
+		// list of namespaces for the cluster
+		nsWithMetricsAccess := []string{}
```
Could a set be used instead for better efficiency? You can convert it to a slice before returning the value in this method.

ckandag replied:
is there a better way to convert  map keys to  a slice  than ranging through the keys and building the slice ourselves?

mprahl replied:
There isn't, but you could consider using this popular library:
https://github.com/deckarep/golang-set

I haven't used it personally though.

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+// this to allow for the cases where only a  kubeClient is available
+// or when rbac api calls are going to be made only for a single user or serviceaccount throughout.
+
+type APIEngine struct {
```
Optional: This name seems rather generic. Could we use a name that better reflects the scope? Perhaps `AccessReviewer` or similar?

ckandag replied:
@mprahl : AccessReviewer.getMetricsAcess() has too much Access in it.   Any other suggestions ?

mprahl replied:
`ACLRetriever`?

ckandag replied:
Changing it to AccessReviewer

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+// When the user's bearer token is used to make the call instead of  impersonation,
+// the user's usergroup are already taken into account when returning the list of acls
+// if the call is successful then this function returns a list of resourcerules
+func MakeSubjectAccessRulesReviewForUser(
```
I think this should be a private function since it's an implementation detail to the API.

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+// for a given cluster,
+//			if all access is configured for the user, i.e the verb "*",  list of the namespaces is set to *
+
+func (r *APIEngine) GetMetricsAccess(userToken string, clusters ...string) (map[string][]string, error) {
```
Can you document either here or in the read me what constitutes metrics access?

mprahl commented on build/common/config/.golangci.yml:
```diff
@@ -0,0 +1,227 @@
+    check-shadowing: false
+  gci:
+    sections:
+      - prefix(github.com/stolostron/go-template-utils)
```
Could you please fix this import path?

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+
+	resourceAccessResults := make(map[string][]string)
+	// search through all the resource rules
+	for _, rule := range resourceRules {
```
You may consider moving this filtering logic to a private helper function so that you can test different permutations rules without having to create the state in Kubernetes.

ckandag replied:
@mprahl , yes, that  is a good point.  Will consider that for the next pass if/when  I have other changes in that code path or  need to update test-cases.

As I have already setup a bunch of test cases. Would like to skip this for now  unless this is abs required for this PR.

mprahl replied:
Not required, but just a suggestion.

mprahl commented on pkg/rbac/rbac_suite_test.go:
```diff
@@ -0,0 +1,254 @@
+	return testUserKClient, nil
+}
+
+func SetupRbacResources(resourcesYaml string) error {
```
To follow Go naming conventions of capitalizing well known acronyms.
```suggestion
func SetupRBACResources(resourcesYaml string) error {
```

mprahl commented:
@ckandag in general, could you please try cleaning up the wording and formatting in the documentation of public methods and functions since that will be public facing to other teams?

gparvin commented on README.md:
```diff
@@ -1,2 +1,2 @@
 # rbac-api-utils
-A library for multi cluster rbac api
+A library for fine-grained multi cluster rbac api
```
Are there any usage scenarios that could be provided beyond just letting developers know they need to run `make test`.  So something like:

This is an example spec of a `Role` and `RoleBinding` object that can be used to provide fine grained access to user `fred` for managed cluster `dev-cluster` and namespace `app-for-fred`.

```
role and role binding here
```

ckandag replied:
@gparvin  I added doc for the function inline with details of the input and out params.
I also added an example scenario with role, role bindings and results for sample invocations of the API
Could you pls take a look again

gparvin replied:
Thanks!

gparvin commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,287 @@
+	response, err := kclient.AuthorizationV1().SelfSubjectRulesReviews().Create(
+		context.TODO(), sarr, metav1.CreateOptions{})
+	if err != nil {
+		log.Fatal(err)
```
Shouldn't we just return the error here?

ckandag replied:
yes, should return err, fixed it thanks

ckandag commented:
@mprahl : Could you also pls re-review the PR

Also, Created this https://issues.redhat.com/browse/ACM-3039 to address the "check for ManagedCluster  access" comment.
As mentioned the obs-rbac already does check for user's access to a ManagedCluster.   So the issue is to determine if it should be left there as is  in obs-rbac  or  atleast partially  consolidated into the new Metrics API (only "get" check on ManagedCluster resource but not the older style Project check)

gparvin commented:
/hold

gparvin commented:
/approve

mprahl commented:
@ckandag could you please conform your commit message to the standards listed here?
https://github.com/open-cluster-management-io/community/blob/main/sig-policy/contribution-guidelines.md#commit-messages

In particular, this guideline:
https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#use-imperative-mood-in-your-commit-message-subject

So for example, your commit message subject (first line) could be `Add the RBAC API for metrics access`.

mprahl commented on README.md:
```diff
@@ -1,2 +1,78 @@
+this library provides utiilies and  wrappers around the kubernetes RBAC API for convenience and 
+to achieve finer grained access control.
+
+Go to the [Contributing guide](CONTRIBUTING.md) to learn how to get involved.
```
Did you commit `CONTRIBUTING.md`?

mprahl commented on README.md:
```diff
@@ -1,2 +1,78 @@
+
+Go to the [Contributing guide](CONTRIBUTING.md) to learn how to get involved.
+
+- The `rbac-api-utils` repository is part of the `open-cluster-management`
```
I'm not sure that this is true since it's in Stolostron.

ckandag replied:
I wondered about this but non OCM component Repos like Cert, Iam or Search etc  all seemed to have the same text.
I will just remove the entire statement

mprahl commented on README.md:
```diff
@@ -1,2 +1,78 @@
+[![License](https://img.shields.io/:license-apache-blue.svg)](http://www.apache.org/licenses/LICENSE-2.0.html)
+
+A library for multi-cluster RBAC API. Access control in OCM is based on RBAC contructs  provided by Kubernetes,
+this library provides utiilies and  wrappers around the kubernetes RBAC API for convenience and 
```
```suggestion
this library provides utiilies and  wrappers around the Kubernetes RBAC API for convenience and 
```

mprahl commented on README.md:
````diff
@@ -1,2 +1,78 @@
+
+```
+
+See [here](./pkg/rbac/rbac.go/#L49) for more information on the parameters for creation of a AccessReviewer
````
```suggestion
See [here](./pkg/rbac/rbac.go/#L49) for more information on the parameters for creation of an `AccessReviewer`.
```

mprahl commented on README.md:
```diff
@@ -1,2 +1,78 @@
+
+### Supported API
+
+**GetMetricsAccess** returns the  managedclusters and namespaces on the managedclusters for which the user has access to view observability metrics. See [here](./pkg/rbac/rbac.go/#L121) for details on the input parameters and results.
```
I think we would want to use the English term of `managed clusters` through this section since you aren't talking about the Kubernetes API objects.

mprahl commented on README.md:
```diff
@@ -1,2 +1,78 @@
+
+OCM Observability gathers  metrics from the managed clusters and stores them for viewing on the Hub. Users can be given access to view metrics for specific namespaces on specific managedclusters.
+
+Consider the scenario where an Application "Blue" is deloyed to namespaces blue1 and blue2 on ManagedClusters  devcluster1 and devcluster2. Inorder to give the Blue admins  access to view Blue metrics, the following clusterroles are set
```
```suggestion
Consider the scenario where an application "Blue" is deployed to namespaces blue1 and blue2 on managed clusters  devcluster1 and devcluster2. In order to give the Blue admins  access to view Blue metrics, the following cluster roles are set
```

mprahl commented on README.md:
````diff
@@ -1,2 +1,78 @@
+
+Consider the scenario where an Application "Blue" is deloyed to namespaces blue1 and blue2 on ManagedClusters  devcluster1 and devcluster2. Inorder to give the Blue admins  access to view Blue metrics, the following clusterroles are set
+
+```
````
```suggestion
```yaml
```

mprahl commented on go.mod:
```diff
@@ -0,0 +1,54 @@
+	golang.org/x/exp v0.0.0-20230108222341-4b8118a2686a
+	k8s.io/api v0.26.0
+	k8s.io/apimachinery v0.26.0
+	k8s.io/client-go v0.26.0
```
You may want to use lower versions since Observability may not be using v0.26.0 yet and this would force them to upgrade.

One reason not to upgrade to newer versions of these Kubernetes modules is that it removes the option for `configmap` leader election which is required for OCP 3.11. This is not applicable for Observability since it's on the Hub, but there may be other reasons.

ckandag replied:
Obs is at v0.25.2 .. downgraded this to v0.25.2 as well

mprahl commented on go.mod:
```diff
@@ -0,0 +1,54 @@
+	k8s.io/apimachinery v0.26.0
+	k8s.io/client-go v0.26.0
+	k8s.io/klog v1.0.0
+	sigs.k8s.io/controller-runtime v0.14.1
```
Same comment as above.

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+// Package rbac provides various Access Review API
```
```suggestion
// Package rbac provides various Access Review APIs
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+)
+
+// ACLConfig holds the  access control configuration  needed to  perform an action
+// e.g. "get" permission on the  "managedclusters" resource is needed to "view" a managedcluster on the ACM Hub
```
```suggestion
// e.g. "get" permission on the  "managedclusters" resource is needed to "view" a managedcluster on the ACM Hub.
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+	verb: "metrics/",
+}
+
+// AccessReviewer is the  API for custom finer-grained access control, it holds the
```
Not sure if you meant this instead.
```suggestion
// AccessReviewer is the  API for custom fined-grained access control, it holds the
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+// It takes two parameters kConfig and kClient, but expects a value to be set for only one of them.
+// An error will be thrown if neither or both values are set.
+//
+// - kConfig is k8s cluster configuration. This should be set when API consumer intents to use
```
```suggestion
// - kConfig is k8s cluster configuration. This should be set when the API consumer intends to use
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+// An error will be thrown if neither or both values are set.
+//
+// - kConfig is k8s cluster configuration. This should be set when API consumer intents to use
+// the AccessReviewer instance to retrieve acls for different users. User specific details( i.e Token)
```
```suggestion
// the AccessReviewer instance to retrieve ACLs for different users. User specific details (i.e Token)
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+// would need to be passed to every invocation of access review API, which will be used to create a
+// k8s client to connect to the cluster to retrieve user specific ACLs.
+//
+// - kClient is K8s cluster client. This should be set when API consumer intents to use
```
```suggestion
// - kClient is K8s cluster client. This should be set when API consumer intends to use
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+	}
+
+	if kClient != nil && kConfig != nil {
+		return nil, errors.New("only one  of either kubeConfig or kubeClient must be a non-nil value")
```
```suggestion
		return nil, errors.New("only one of either kubeConfig or kubeClient must be a non-nil value")
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+
+// GetMetricsAccess retrieves the user's ACLs from the k8s cluster  and processes them to determine
+// user's access to observability metrics that are gathered from managed clusters.
+// It returns a map of allowed managedclusters and the allowed namespaces for each managedcluster.
```
```suggestion
// It returns a map where the keys are managed clusters and the values are slices of allowed namespaces.
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+//
+// - userToken is the user's OAuth bearer token, is required if k8s config was set on the AccessReviewer
+//
+// - clusters are the  names of the managedclusters for which  allowed metrics access is processed and returned.
```
```suggestion
// - clusters are the  names of the managed clusters for which allowed metrics access is returned.
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+// - userToken is the user's OAuth bearer token, is required if k8s config was set on the AccessReviewer
+//
+// - clusters are the  names of the managedclusters for which  allowed metrics access is processed and returned.
+// if no clusters are specified, then  metrics access is returned for all "allowed" managedclusters.
```
```suggestion
// If no clusters are specified, then metrics access is returned for all "allowed" managed clusters.
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+func (r *AccessReviewer) GetMetricsAccess(userToken string, clusters ...string) (map[string][]string, error) {
+	klog.V(2).Infof("GetMetricsAccess for clusters: %v", clusters)
+
+	// get Client to talk to the Kubernetes cluster
```
Optional: In general, comments that are pointing out the obvious aren't a best practice since they don't provide value and they can get stale over time and become inaccurate.

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+			}
+		}
+
+		nsWithMetricsAccess := []string{}
```
Optional: This is a good practice to avoid `append` unnecessarily recreating the underlying array as it grows.
```suggestion
		nsWithMetricsAccess := make([]string{}, 0, len(metricsAccessMap))
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+		klog.V(2).Infof("clustername [%s] acls[%s]\n", clustername, clusteracls)
+
+		// list of namespaces for the cluster
+		metricsAccessMap := make(map[string]bool)
```
Optional:
```suggestion
		metricsAccessMap := make(map[string]bool, len(clusteracls))
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+	klog.V(2).Infof(" resource access results: %v", resourceACLs)
+
+	// from the list of all ACLs for ManagedCluster, filter out the "metrics" specific acls and grab the namespaces
+	metricsAccessResults := make(map[string][]string)
```
Optional:
```suggestion
	metricsAccessResults := make(map[string][]string, len(resourcACLs))
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+	return metricsAccessResults, nil
+}
+
+// GetResourceAccess returns all configured ACLs for a given resource type
```
These periods are important for how they will be shown in pkg.go.dev.
```suggestion
// GetResourceAccess returns all configured ACLs for a given resource type.
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+
+// GetResourceAccess returns all configured ACLs for a given resource type
+// It returns a map of resource names and ACLs for that resource. for a given resource,
+// if no  acls are configured, an empty list is returned for it in the results.
```
```suggestion
// if no ACLs are configured, an empty slice is returned for it in the results.
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+// It returns a map of resource names and ACLs for that resource. for a given resource,
+// if no  acls are configured, an empty list is returned for it in the results.
+//
+// - resourcenames are the names of the resources for which acls are to processed and returned,
```
```suggestion
// - resourcenames are the names of the resources for which ACLs are returned,
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+// if no  acls are configured, an empty list is returned for it in the results.
+//
+// - resourcenames are the names of the resources for which acls are to processed and returned,
+// if no resource names are passed,acls for all allowed resources of the given type are processed and returned.
```
```suggestion
// if no resource names are passed, ACLs for all allowed resources of the given type are returned.
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+	return resourceAccessResults, nil
+}
+
+// addUniqueItems a convenience method for building a list with unique entries
```
```suggestion
// addUniqueItems a convenience method for building a slice with unique entries
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+		}
+	}
+
+	klog.V(2).Infof("Result List: %v ", resultList)
```
This is likely too noisy. I'd even consider removing this log message.
```suggestion
	klog.V(3).Infof("Result List: %v ", resultList)
```

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,314 @@
+// - resourcenames are the names of the resources for which acls are to processed and returned,
+// if no resource names are passed,acls for all allowed resources of the given type are processed and returned.
+//
+// - namespace is used for namespace-scoped resources, for cluster-scoped resources it should be left empty
```
```suggestion
// - namespace is used for namespace-scoped resources, for cluster-scoped resources it should be left empty. If not specified, it defaults to the value "default" for namespace-scoped resources.
```

mprahl reviewed (commented):
@ckandag this looks great! Most of my comments are documentation related.

After that's addressed, I'll approve it.

ckandag commented:
> @ckandag this looks great! Most of my comments are documentation related.
> 
> After that's addressed, I'll approve it.

@mprahl Could you pls take a look at it again, I moved the versions in go.mod to  match current versions in Obs Repo and also addressed all the doc'ing errors

gparvin reviewed (approved):
It looks good to me!  Thanks Chaitanya!
I put a hold in place for Matt and Dale to finish reviewing.

mprahl commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,54 @@
+**Table of Contents**
```
Could you please use the same style as this?
https://github.com/stolostron/config-policy-controller/blob/main/CONTRIBUTING.md

mprahl commented on pkg/rbac/rbac.go:
```diff
@@ -0,0 +1,315 @@
+// if no  ACLs are configured, an empty list is returned for it in the results.
+//
+// - resourcenames are the names of the resources for which ACLs are and returned,
+// if no resource names are passed,ACLs for all allowed resources of the given type are returned.
```
```suggestion
// if no resource names are passed, ACLs for all allowed resources of the given type are returned.
```

ckandag commented:
/unhold

ckandag commented:
/approve

mprahl commented:
@ckandag I think you'll have to manually merge it this time since there is no OWNERS file in the repo yet unless you added an OWNERS file in Prow already.

ckandag commented:
> @ckandag I think you'll have to manually merge it this time since there is no OWNERS file in the repo yet unless you added an OWNERS file in Prow already.

I see, ty!
